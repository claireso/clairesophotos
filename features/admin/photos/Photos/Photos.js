import { Fragment, useEffect, useCallback } from 'react'
import { useRouter } from 'next/router'

import PhotosReducer, { ACTION_TYPES } from '../reducer'

import Loader from '@components/Loader'
import { IconPlus } from '@components/Icons'
import { List, ListHeader } from '@components/List'
import { PrimaryButton } from '@components/Buttons'
import { Heading1 } from '@components/Headings'
import Pager from '@components/Pager'

import Photo from './components/Photo'

const Photos = () => {
  const [
    { items: photos, pager, ...state },
    { loadResources }
  ] = PhotosReducer.usePhotosReducer()
  const isLoading = ['idle', 'loading'].includes(state.status)

  const router = useRouter()
  const {
    query: { page },
    pathname
  } = router

  useEffect(() => {
    loadResources(page)
  }, [page, loadResources])

  const navigate = useCallback(
    (params) => {
      const query = {}
      if (page) query.page = page
      router.push({ pathname: pathname, query: { ...query, ...params } })
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [page]
  )

  const onPageChange = (page) => navigate({ page })

  const onCreate = useCallback(
    (ev) => {
      ev.preventDefault()
      navigate({ action: ACTION_TYPES.CREATE })
    },
    [navigate]
  )

  const onDelete = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.DELETE,
        id: id
      })
    },
    [navigate]
  )

  const onEdit = useCallback(
    (id, event) => {
      event.preventDefault()
      navigate({
        action: ACTION_TYPES.EDIT,
        id: id
      })
    },
    [navigate]
  )

  return (
    <Fragment>
      <ListHeader>
        <Heading1 data-testid="list-heading">
          Your photos {pager && <span>({pager.count})</span>}
        </Heading1>
        <PrimaryButton data-testid="button-create" onClick={onCreate}>
          Add a new photo
          <IconPlus />
        </PrimaryButton>
      </ListHeader>

      {isLoading ? (
        <Loader />
      ) : (
        <Fragment>
          <List>
            {photos.map((photo, index) => (
              <Photo
                key={index}
                {...photo}
                onDelete={onDelete}
                onEdit={onEdit}
              />
            ))}
          </List>
          <Pager {...pager} navigate={onPageChange} />
        </Fragment>
      )}
    </Fragment>
  )
}

export default Photos
