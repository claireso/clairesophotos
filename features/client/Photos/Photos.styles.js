import styled from 'styled-components'

export const PagerWrapper = styled.div`
  grid-column: 1 / -1;
  margin: 4.5rem 0;

  @media (min-width: 800px) {
    margin: 8.5rem 0;
  }
`
