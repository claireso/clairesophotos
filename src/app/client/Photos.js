import React from 'react'
import PropTypes from 'prop-types'

import Photo from './components/Photo'
import Pager from './components/Pager'
import Button from './components/Button'

const Photos = (props = {}) => {
  return (
    <React.Fragment>
      {props.photos.map((photo, index) => <Photo key={index} {...photo} />)}
      <Pager
        {...props.pager}
        navigate={page => {
          const state = { page }
          window.history.pushState(state, '', `?page=${page}`)

          const popStateEvent = new PopStateEvent('popstate', { state })
          dispatchEvent(popStateEvent)
        }}
      >
        {({ items, getItemsProps }) => {
          return items.map((item, key) => (
            <li key={key}>
              <Button
                {...getItemsProps({
                  label: item.label,
                  title: item.title,
                  item: item
                })}
              />
            </li>
          ))
        }}
      </Pager>
    </React.Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
