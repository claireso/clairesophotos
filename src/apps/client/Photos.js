import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import Photo from './components/Photo'
import Pager from './components/Pager'

const Photos = (props = {}) => {
  const navigate = useCallback(page => {
    const state = { page }
    window.history.pushState(state, '', `?page=${page}`)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }, [])

  return (
    <React.Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} />
      ))}
      <Pager {...props.pager} navigate={navigate} />
    </React.Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
