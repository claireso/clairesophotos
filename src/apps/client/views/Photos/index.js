import React, { Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'

import Photo from '../../components/Photo'
import Pager from '../../components/Pager'

const Photos = (props = {}) => {
  const navigate = useCallback(page => {
    const state = { page }
    window.history.pushState(state, '', `?page=${page}`)

    const popStateEvent = new PopStateEvent('popstate', { state })
    dispatchEvent(popStateEvent)
  }, [])

  return (
    <Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} row={index + 1} />
      ))}
      <Pager {...props.pager} navigate={navigate} />
    </Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
