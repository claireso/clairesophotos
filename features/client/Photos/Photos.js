import { Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'
import Router from 'next/router'

import Pager from '@components/Pager'

import Photo from './components/Photo'

import * as S from './Photos.styles'

const Photos = (props = {}) => {
  const navigate = useCallback((page) => {
    Router.push({ pathname: '/', query: { page } })
    window.scrollTo(0, 0)
  }, [])

  return (
    <Fragment>
      {props.photos.map((photo, index) => (
        <Photo key={index} {...photo} row={index + 1} />
      ))}
      <S.PagerWrapper>
        <Pager {...props.pager} navigate={navigate} />
      </S.PagerWrapper>
    </Fragment>
  )
}

Photos.propTypes = {
  photos: PropTypes.array.isRequired,
  pager: PropTypes.object.isRequired
}

export default Photos
