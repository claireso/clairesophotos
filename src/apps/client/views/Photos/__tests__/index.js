import React from 'react'
import { render, fireEvent } from '@testing-library/react'

import Photos from '../index'

describe('<Photos />', () => {
  test('should render component', () => {
    const { items: photos, pager } = global.__PHOTOS__
    const { container } = render(<Photos photos={photos} pager={pager} />)

    expect(container).toMatchSnapshot()
  })

  test('should call `pushState` and `popState`', () => {
    window.history.pushState = jest.fn()
    window.onpopstate = jest.fn()

    const { items: photos, pager } = global.__PHOTOS__
    const { getByTitle } = render(<Photos photos={photos} pager={pager} />)

    fireEvent.click(getByTitle('Next page'))

    expect(window.history.pushState).toHaveBeenCalledWith(
      { page: 2 },
      '',
      '?page=2'
    )
    expect(window.onpopstate).toHaveBeenCalledTimes(1)

    fireEvent.click(getByTitle('Last page'))

    expect(window.history.pushState).toHaveBeenCalledWith(
      { page: 19 },
      '',
      '?page=19'
    )
    expect(window.onpopstate).toHaveBeenCalledTimes(2)
  })
})
