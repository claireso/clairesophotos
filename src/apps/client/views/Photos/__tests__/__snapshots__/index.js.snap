// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Photos /> should render component 1`] = `
.c0 {
  grid-row-start: 1;
  grid-column: 1 / span 6;
  margin: 0;
}

.c4 {
  grid-row-start: 2;
  grid-column: 1 / span 6;
  margin: 0;
}

.c2 {
  font-size: 1.2rem;
}

.c3 {
  color: var(--gray-2);
  display: block;
  font-size: 1rem;
  font-style: italic;
  margin: 0.4rem 0 0;
}

.c1 {
  background: var(--gray-4);
  position: relative;
  padding-top: calc(0.666667 * 100%);
  width: 100%;
  margin: 0 0 1rem;
}

.c6 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  cursor: pointer;
  outline: none;
  padding: 1rem 1.2rem 1.1rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: white;
  border: 1px solid var(--gray-1);
  display: inline-block;
  font-family: 'Roboto',Arial,sans-serif;
  width: 4.2rem;
  color: var(--text);
  font-size: 1.6rem;
  font-weight: 700;
  will-change: border-color;
  -webkit-transition: border-color 150ms ease-out;
  transition: border-color 150ms ease-out;
}

.c6:hover {
  border-color: var(--gray-3);
}

.c5 {
  grid-column: 1 / -1;
  margin: 0;
  padding: 0;
  text-align: center;
}

.c5 > li {
  display: inline-block;
  margin: 0 0.5rem;
}

@media (min-width:800px) {
  .c0 {
    grid-column: calc(var(--grid-number-column-large) - 8) / span 8;
  }

  .c0 + figure {
    margin-top: 20rem;
  }
}

@media (min-width:800px) {
  .c4 {
    grid-column: 3 / span 8;
  }

  .c4 + figure {
    margin-top: 20rem;
  }
}

@media (min-width:800px) {
  .c5 {
    margin: 8.5rem 0;
  }
}

<div>
  <figure
    class="c0"
  >
    <div
      class="c1"
    />
    <figcaption
      class="c2"
    >
      <span />
      <span
        class="c3"
      >
        Février 2019
      </span>
    </figcaption>
  </figure>
  <figure
    class="c4"
  >
    <div
      class="c1"
    />
    <figcaption
      class="c2"
    >
      <span />
      <span
        class="c3"
      >
        Janvier 2019
      </span>
    </figcaption>
  </figure>
  <ul
    class="c5"
  >
    <li>
      <button
        class="sc-AykKH c6"
        title="Next page"
      >
        »
      </button>
    </li>
    <li>
      <button
        class="sc-AykKH c6"
        title="Last page"
      >
        »»
      </button>
    </li>
  </ul>
</div>
`;
